kpi-tracker/src/main.rs

```diff
@@ -965,20 +1104,75 @@ async fn main() -> anyhow::Result<()> {
     // Create a channel between the task querying the node and the task logging
     // transactions.
     let (block_sender, block_receiver) = tokio::sync::mpsc::channel(100);
+    // node/db processes run until the stop flag is triggered.
+    let stop_flag = Arc::new(AtomicBool::new(false));
+    let shutdown_handle = tokio::spawn(set_shutdown(stop_flag.clone()));
 
-    tokio::spawn(async {
-        if let Err(e) = run_db_process(args.db_connection, block_receiver, height_sender).await {
-            log::error!("Error happened while running DB process: {}", e);
-        }
-    });
+    let db_handle = {
+        let stop_flag = stop_flag.clone();
+        tokio::spawn(async move {
+            if let Err(e) =
+                run_db_process(args.db_connection, block_receiver, height_sender, stop_flag).await
+            {
+                log::error!("Error happened while running DB process: {}", e);
+            }
+        })
+    };
 
-    let latest_height = height_receiver
+    let mut latest_height = height_receiver
         .await
         .context("Did not receive height of most recent block recorded in database")?;
 
-    node_process(args.node, latest_height, block_sender)
-        .await
-        .context("Error happened while querying node.")?;
+    let mut latest_successful_node: u64 = 0;
+    let num_nodes = args.node_endpoints.len() as u64;
+    for (node, i) in args.node_endpoints.into_iter().cycle().zip(0u64..) {
+        let start_height = latest_height;
+        if stop_flag.load(Ordering::Acquire) || block_sender.is_closed() {
```

Comment:
Can you elaborate on what the "block_sender.is_closed()" is there for?

Reply:
If database connection was lost and couldn't be re-established, this shuts the service down. I added an inline comment explaining why it's included in the conditional now.

The DB process already logs an error in this case before shutting down, so my thought is that would be redundant to also add special handling for this case here. What do you think?

Reply:
The error is logged, but the service won't stop without this check.

You do things a bit differently than in the transaction logger. There we return Ok(()) if database stops, but here you instead return Err from he node_process.

That then means that this loop will not terminate properly without the `|| block_sender...`.

I would recommend you do the same as we do in the transaction logger. Otherwise we do need the check here.

Reply:
You're right. Might as well return directly from the process handling the DB stuff, is much easier to follow.